hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/placement/algorithms/SCMContainerPlacementRackScatter.java

```diff
@@ -0,0 +1,343 @@
+          Node node = chooseNode(unavailableNodes, affinityNode,
+              metadataSizeRequired, dataSizeRequired);
+          if (node != null) {
+            chosenNodes.add(node);
+            mutableFavoredNodes.remove(node);
+            unavailableNodes.add(node);
+            nodesRequired--;
+            if (nodesRequired == 0) {
+              break;
+            }
+          }
+        }
+        chosenRacksInForLoop.add(rack);
+      }
+      toChooseRack.removeAll(chosenRacksInForLoop);
+      if (chosenListSize == chosenNodes.size()) {
+        retryCount++;
+      } else {
+        retryCount = 0;
+      }
+    }
+
+    return Arrays.asList(chosenNodes.toArray(new DatanodeDetails[0]));
+  }
+
+  @Override
+  public DatanodeDetails chooseNode(List<DatanodeDetails> healthyNodes) {
+    return null;
+  }
+
+  /**
+   * Choose a datanode which meets the requirements. If there is no node which
+   * meets all the requirements, there is fallback chosen process depending on
+   * whether fallback is allowed when this class is instantiated.
+   *
+   *
+   * @param excludedNodes - list of the datanodes to excluded. Can be null.
+   * @param affinityNode - the chosen nodes should be on the same rack as
+   *                    affinityNode. Can be null.
+   * @param dataSizeRequired - size required for the container.
+   * @param metadataSizeRequired - size required for Ratis metadata.
+   * @return List of chosen datanodes.
+   * @throws SCMException  SCMException
+   */
+  private Node chooseNode(List<Node> excludedNodes, Node affinityNode,
+      long metadataSizeRequired, long dataSizeRequired) throws SCMException {
+    int ancestorGen = RACK_LEVEL;
+    int maxRetry = MAX_RETRY;
+    List<String> excludedNodesForCapacity = null;
+    while(true) {
+      metrics.incrDatanodeChooseAttemptCount();
+      Node node = networkTopology.chooseRandom(NetConstants.ROOT,
+          excludedNodesForCapacity, excludedNodes, affinityNode, ancestorGen);
+      if (node == null) {
+        // cannot find the node which meets all constrains
+        LOG.warn("Failed to find the datanode for container. excludedNodes:" +
+            (excludedNodes == null ? "" : excludedNodes.toString()) +
+            ", affinityNode:" +
+            (affinityNode == null ? "" : affinityNode.getNetworkFullPath()));
+        return null;
+      }
+
+      DatanodeDetails datanodeDetails = (DatanodeDetails)node;
+      if (isValidNode(datanodeDetails, metadataSizeRequired,
+          dataSizeRequired)) {
+        return node;
+      }
+
+      maxRetry--;
+      if (maxRetry == 0) {
+        // avoid the infinite loop
+        String errMsg = "No satisfied datanode to meet the space constrains. "
+            + "metadatadata size required: " + metadataSizeRequired +
+            " data size required: " + dataSizeRequired;
+        LOG.info(errMsg);
+        return null;
+      }
+      if (excludedNodesForCapacity == null) {
```

Comment:
Should a node which is fails the `isValidNode()` check, be added to the excludedNode list too? This means that node will not be retried again if this rack is considered again in a retry?

Lets say we have a rack with 10 nodes, but 8 are over capacity. 2 are good. There is a fairly good chance that in 3 retries, we will not find either of the good nodes. Then the outer loop in `chooseDatanodes` may consider this rack again, and it could fall into the same problem. However if we add the nodes to the exclude list, that list will be passed back into this method on a retry in the outer loop and we have a better chance of finding the nodes with capacity on the second or third pass.